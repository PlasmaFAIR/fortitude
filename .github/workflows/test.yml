name: Tests

on:
  push:
    paths:
      - '**.rs'
      - '**/Cargo.toml'
      - '**/Cargo.lock'
      - '.github/workflows/test.yml'
  pull_request:
    paths:
      - '**.rs'
      - '**/Cargo.toml'
      - '**/Cargo.lock'
      - '.github/workflows/test.yml'
  workflow_dispatch:

env:
  CARGO_INCREMENTAL: 0
  CARGO_TERM_COLOR: always

jobs:

  build-linux:
    name: Build and Test
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v4
    - name: "Install cargo nextest"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-nextest
    - name: "Install cargo insta"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-insta
    - name: "Build"
      run: cargo build --verbose
    - name: "Run tests"
      shell: bash
      env:
        NEXTEST_PROFILE: "ci"
      run: cargo insta test --all-features --unreferenced reject --test-runner nextest
    - name: "Check docs"
      run: cargo dev generate-all --mode=check
    - name: "Check docs for broken links"
      run: cargo doc --all --no-deps
      env:
        # Setting RUSTDOCFLAGS because `cargo doc --check` isn't yet
        # implemented (https://github.com/rust-lang/cargo/issues/10025)
        RUSTDOCFLAGS: "-D warnings"
    - uses: actions/upload-artifact@v4
      with:
        name: fortitude
        path: target/debug/fortitude

  build-windows:
    name: Build and Test (windows)
    runs-on: windows-latest
    steps:
    - uses: actions/checkout@v4
    - name: "Install cargo nextest"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-nextest
    - name: "Install cargo insta"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-insta
    - name: "Build"
      run: cargo build --verbose
    - name: "Run tests"
      shell: bash
      env:
        NEXTEST_PROFILE: "ci"
      run: cargo insta test --all-features --unreferenced reject --test-runner nextest

  build-macos:
    name: Build and Test (macos)
    runs-on: macos-latest
    steps:
    - uses: actions/checkout@v4
    - name: "Install cargo nextest"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-nextest
    - name: "Install cargo insta"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-insta
    - name: "Build"
      run: cargo build --verbose
    - name: "Run tests"
      shell: bash
      env:
        NEXTEST_PROFILE: "ci"
      run: cargo insta test --all-features --unreferenced reject --test-runner nextest

  cargo-install:
    name: Test cargo installation
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v4
    - name: Check cargo install
      run: |
        cargo install --path fortitude
        fortitude --help

  pip-install:
    name: Test pip installation
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v4
    - name: Set up Python
      uses: actions/setup-python@v5
      with:
        python-version: '3.x'
    - name: Check pip install
      run: |
        pip install .
        fortitude --help
