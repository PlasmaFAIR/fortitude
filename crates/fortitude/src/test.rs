use std::path::Path;

use anyhow::Result;
use ruff_source_file::{SourceFile, SourceFileBuilder};

use crate::{
    check::{
        ast_entrypoint_map, check_file, read_to_string, rules_to_path_rules, rules_to_text_rules,
    },
    message::{Emitter, TextEmitter},
    rule_table::RuleTable,
    rules::Rule,
    settings::{self, FixMode, Settings},
};

#[macro_export]
macro_rules! apply_common_filters {
    {} => {
        let mut settings = insta::Settings::clone_current();
        // Convert windows paths to Unix Paths.
        settings.add_filter(r"\\\\?([\w\d.])", "/$1");
        let _bound = settings.bind_to_scope();
    }
}

pub(crate) fn test_resource_path(path: impl AsRef<Path>) -> std::path::PathBuf {
    Path::new("./resources/test/").join(path)
}

/// Run [`check_path`] on a Fortran file in the `resources/test/fixtures` directory.
pub(crate) fn test_path(
    path: impl AsRef<Path>,
    rules: &[Rule],
    settings: &Settings,
) -> Result<String> {
    let path = test_resource_path("fixtures").join(path);
    let source = read_to_string(&path)?;
    let filename = path.to_string_lossy();
    let file = SourceFileBuilder::new(filename.as_ref(), source.as_str()).finish();
    Ok(test_contents(&path, &file, rules, settings))
}

pub(crate) fn test_contents(
    path: &Path,
    file: &SourceFile,
    rules: &[Rule],
    settings: &Settings,
) -> String {
    let rule_table = RuleTable::from_iter(rules.iter().cloned());
    let path_rules = rules_to_path_rules(&rule_table);
    let text_rules = rules_to_text_rules(&rule_table);
    let ast_entrypoints = ast_entrypoint_map(&rule_table);

    match check_file(
        &rule_table,
        &path_rules,
        &text_rules,
        &ast_entrypoints,
        path,
        file,
        settings,
        FixMode::Generate,
        settings::IgnoreAllowComments::Disabled,
    ) {
        Ok(violations) => {
            if violations.messages.is_empty() {
                return String::new();
            }

            let mut output = Vec::new();

            TextEmitter::default()
                .with_show_fix_status(true)
                .with_show_fix_diff(true)
                .with_show_source(true)
                .with_unsafe_fixes(crate::settings::UnsafeFixes::Enabled)
                .emit(&mut output, &violations.messages)
                .unwrap();

            String::from_utf8(output).unwrap()
        }
        Err(msg) => {
            panic!("Failed to process: {msg}");
        }
    }
}
