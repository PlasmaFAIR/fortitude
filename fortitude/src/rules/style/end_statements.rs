use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Fix};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What does it do?
/// Checks that `end` statements include the type of construct they're ending
///
/// ## Why is this bad?
/// End statements should specify what kind of construct they're ending, and the
/// name of that construct. For example, prefer this:
///
/// ```f90
/// module mymodule
///   ...
/// end module mymodule
/// ```
///
/// To this:
///
/// ```f90
/// module mymodule
///   ...
/// end
/// ```
///
/// Or this:
///
/// ```f90
/// module mymodule
///   ...
/// end module
/// ```
///
/// Similar rules apply for many other Fortran statements
#[derive(ViolationMetadata)]
pub(crate) struct UnnamedEndStatement {
    statement: String,
    name: String,
}

impl AlwaysFixableViolation for UnnamedEndStatement {
    #[derive_message_formats]
    fn message(&self) -> String {
        "end statement should be named.".to_string()
    }

    fn fix_title(&self) -> String {
        let Self { statement, name } = self;
        format!("Write as 'end {statement} {name}'.")
    }
}

/// Maps declaration kinds to its name and the kind of the declaration statement node
fn map_declaration(kind: &str) -> (&'static str, &'static str) {
    match kind {
        "module" => ("module", "module_statement"),
        "submodule" => ("submodule", "submodule_statement"),
        "program" => ("program", "program_statement"),
        "function" => ("function", "function_statement"),
        "subroutine" => ("subroutine", "subroutine_statement"),
        "module_procedure" => ("procedure", "module_procedure_statement"),
        "derived_type_definition" => ("type", "derived_type_statement"),
        _ => unreachable!("Invalid entrypoint for AbbreviatedEndStatement"),
    }
}

impl AstRule for UnnamedEndStatement {
    fn check<'a>(
        _settings: &Settings,
        node: &'a Node,
        src: &'a SourceFile,
    ) -> Option<Vec<Diagnostic>> {
        // TODO Also check for optionally labelled constructs like 'do' or 'select'

        // If end node is named, move on.
        // Not catching incorrect end statement name here, as the compiler should
        // do that for us.
        if node.child_with_name("name").is_some() {
            return None;
        }

        let declaration = node.parent()?;
        let (statement, statement_kind) = map_declaration(declaration.kind());
        let statement_node = declaration.child_with_name(statement_kind)?;
        let name_kind = match statement_kind {
            "derived_type_statement" => "type_name",
            _ => "name",
        };
        let name = statement_node
            .child_with_name(name_kind)?
            .to_text(src.source_text())?
            .to_string();
        let statement = statement.to_string();
        let replacement = format!("end {statement} {name}");
        let fix = Fix::safe_edit(node.edit_replacement(src, replacement));
        some_vec![Diagnostic::from_node(Self { statement, name }, node).with_fix(fix)]
    }

    fn entrypoints() -> Vec<&'static str> {
        vec![
            "end_module_statement",
            "end_submodule_statement",
            "end_program_statement",
            "end_function_statement",
            "end_subroutine_statement",
            "end_module_procedure_statement",
            "end_type_statement",
        ]
    }
}
