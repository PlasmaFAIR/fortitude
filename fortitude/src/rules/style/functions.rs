use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What it does
///
/// Checks that every function explicitly declares which variable contains the result.
///
/// ## Why is this bad?
/// Specifying the `result()` clause allows for the variable inside the function containing the
/// result to be named differently from the function.
///
/// Allowing the variable to be named different from the function allows for better naming of both
/// the function and the internal variables, and also it can help when creating functions that might
/// need to be duplicated to have versions supporting various types. Since declaring the result in
/// the `result()` clause allows for easier duplication of the function because then less lines must
/// be changed.
///
/// ## Example
/// ```f90
/// interface distance
///     module procedure distance_real32, distance_real64
/// end interface distance
///
/// function distance_real32(a, b)
///     real(kind=real32), intent(in) :: a, b
///     real(kind=real32)             :: distance_real32
///
///     distance_real32 = abs(a-b)
/// end function distance_real
///
/// function distance_real64(a, b)
///     real(kind=real64), intent(in) :: a, b
///     real(kind=real64)             :: distance_real64
///
///     distance_real64 = abs(a-b)
/// end function distance_real
/// ```
///
/// Use instead:
/// ```f90
/// interface distance
///     module procedure distance_real32, distance_real64
/// end interface distance
///
/// function distance_real32(a, b) result(distance)
///     integer, parameter               :: func_type = real32
///
///     real(kind=func_type), intent(in) :: a, b
///     real(kind=func_type)             :: distance
///
///     distance = abs(a-b)
/// end function distance_real
///
/// function distance_real64(a, b) result(distance)
///     integer, parameter               :: func_type = real64
///
///     real(kind=func_type), intent(in) :: a, b
///     real(kind=func_type)             :: distance
///
///     distance = abs(a-b)
/// end function distance_real
/// ```
///
/// ## References
/// - Clerman, N. Spector, W., 2012, _Modern Fortran: Style and Usage_, Cambridge
///   University Press, Rule 129, p. 178.
#[derive(ViolationMetadata)]
pub(crate) struct FunctionMissingResult;

impl Violation for FunctionMissingResult {
    #[derive_message_formats]
    fn message(&self) -> String {
        "Function missing result() specifier".to_string()
    }
}

impl AstRule for FunctionMissingResult {
    fn check<'a>(
        _settings: &Settings,
        node: &'a Node,
        _src: &'a SourceFile,
    ) -> Option<Vec<Diagnostic>> {
        // Just need to check for the presence of the function_result node
        if node.child_with_name("function_result").is_some() {
            return None;
        }

        some_vec![Diagnostic::from_node(Self {}, node)]
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["function_statement"]
    }
}
