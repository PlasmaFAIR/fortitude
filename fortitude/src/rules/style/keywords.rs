/// Defines rules that govern the use of keywords.
use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix, FixAvailability, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use ruff_text_size::{TextRange, TextSize};
use std::str::FromStr;
use tree_sitter::Node;

// TODO Support for `endfile`/`end file`

#[derive(strum_macros::EnumString, strum_macros::Display)]
#[strum(serialize_all = "lowercase", ascii_case_insensitive)]
enum DoubleKeyword {
    DoublePrecision,
    DoubleComplex,
    SelectCase,
    SelectType,
    ElseIf,
    ElseWhere,
    EndAssociate,
    EndBlock,
    EndCritical,
    EndDo,
    EndEnum,
    EndForAll,
    EndFunction,
    EndIf,
    EndInterface,
    EndModule,
    EndProcedure,
    EndProgram,
    EndSelect,
    EndSubmodule,
    EndSubroutine,
    EndTeam,
    EndType,
    EndWhere,
    InOut,
    GoTo,
}

impl DoubleKeyword {
    /// Returns the number of characters that the first keyword is comprised of.
    fn offset(&self) -> usize {
        match self {
            DoubleKeyword::InOut | DoubleKeyword::GoTo => 2,
            DoubleKeyword::DoublePrecision
            | DoubleKeyword::DoubleComplex
            | DoubleKeyword::SelectCase
            | DoubleKeyword::SelectType => 6,
            DoubleKeyword::ElseIf | DoubleKeyword::ElseWhere => 4,
            DoubleKeyword::EndAssociate
            | DoubleKeyword::EndBlock
            | DoubleKeyword::EndCritical
            | DoubleKeyword::EndDo
            | DoubleKeyword::EndEnum
            | DoubleKeyword::EndForAll
            | DoubleKeyword::EndFunction
            | DoubleKeyword::EndIf
            | DoubleKeyword::EndInterface
            | DoubleKeyword::EndModule
            | DoubleKeyword::EndProcedure
            | DoubleKeyword::EndProgram
            | DoubleKeyword::EndSelect
            | DoubleKeyword::EndSubmodule
            | DoubleKeyword::EndSubroutine
            | DoubleKeyword::EndTeam
            | DoubleKeyword::EndType
            | DoubleKeyword::EndWhere => 3,
        }
    }

    /// Return keywords with a space between them.
    fn with_space(&self) -> String {
        let mut result = self.to_string();
        result.insert(self.offset(), ' ');
        result
    }
}

/// ## What it does
/// Checks for the use of keywords comprised of two words where the space is
/// omitted, such as `elseif` instead of `else if` and `endmodule` instead of
/// `endmodule`. The keywords `inout` and `goto` are exempt from this rule by
/// default, but may be included by setting the options
/// [`inout_with_space`](../settings.md#inout-with-space) and
/// [`goto_with_space`](../settings.md#goto-with-space).
///
/// ## Why is this bad?
/// Contracting two keywords into one can make code less readable. Enforcing
/// this rule can help maintain a consistent style.
#[derive(ViolationMetadata)]
pub struct KeywordsMissingSpace {
    keywords: DoubleKeyword,
}

impl AlwaysFixableViolation for KeywordsMissingSpace {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { keywords } = self;
        format!("Missing space in '{keywords}'")
    }

    fn fix_title(&self) -> String {
        let preferred = self.keywords.with_space();
        format!("Replace with '{preferred}'")
    }
}

impl AstRule for KeywordsMissingSpace {
    fn check(settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let first_child = if node.kind() == "inout" {
            *node
        } else {
            node.child(0)?
        };
        let text = first_child.to_text(src.source_text())?;
        let keywords = DoubleKeyword::from_str(text).ok()?;

        // Exit early if the keyword is permitted
        if matches!(keywords, DoubleKeyword::InOut)
            && !settings.check.keyword_whitespace.inout_with_space
        {
            return None;
        }
        if matches!(keywords, DoubleKeyword::GoTo)
            && !settings.check.keyword_whitespace.goto_with_space
        {
            return None;
        }

        let space_pos = TextSize::try_from(node.start_byte() + keywords.offset()).unwrap();
        let fix = Fix::safe_edit(Edit::insertion(" ".to_string(), space_pos));
        some_vec!(Diagnostic::from_node(Self { keywords }, &first_child).with_fix(fix))
    }

    fn entrypoints() -> Vec<&'static str> {
        vec![
            "elseif_clause",
            "elsewhere_clause",
            "end_associate_statement",
            "end_block_construct_statement",
            "end_coarray_critical_statement",
            "end_coarray_team_statement",
            "end_do_loop_statement",
            "end_enum_statement",
            "end_forall_statement",
            "end_function_statement",
            "end_if_statement",
            "end_interface_statement",
            "end_module_procedure_statement",
            "end_module_statement",
            "end_program_statement",
            "end_select_statement",
            "end_submodule_statement",
            "end_subroutine_statement",
            "end_type_statement",
            "end_where_statement",
            "inout",
            "intrinsic_type",    // double precision and double complex
            "keyword_statement", // goto
            "select_case_statement",
            "select_type_statement",
        ]
    }
}

/// ## What it does
/// Checks for the use of `in out` instead of `inout` and `go to` instead of `goto`.
/// Either may be exempted from this rule by setting the options
/// [`inout_with_space`](../settings.md#inout-with-space) and
/// [`goto_with_space`](../settings.md#goto-with-space).
///
/// ## Why is this bad?
/// By convention, `inout` in normally preferred to `in out`. Both `go to` and
/// `goto` are valid, but Fortitude prefers the latter as `goto` is most common
/// in other languages, and neither `go` nor `to` have secondary purposes in
/// other keywords. Enforcing this rule can help maintain a consistent style.
#[derive(ViolationMetadata)]
pub struct KeywordHasWhitespace {
    keywords: DoubleKeyword,
}

impl Violation for KeywordHasWhitespace {
    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;

    #[derive_message_formats]
    fn message(&self) -> String {
        let keywords = self.keywords.with_space();
        format!("Whitespace included in '{keywords}'")
    }

    fn fix_title(&self) -> Option<String> {
        let preferred = self.keywords.to_string();
        Some(format!("Replace with '{preferred}'"))
    }
}

impl AstRule for KeywordHasWhitespace {
    fn check(settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        if node.kind() == "inout" && settings.check.keyword_whitespace.inout_with_space {
            return None;
        }
        if node.kind() == "keyword_statement" && settings.check.keyword_whitespace.goto_with_space {
            return None;
        }
        let (first, second, first_child, violation) = if node.kind() == "inout" {
            (
                "in",
                "out",
                *node,
                Self {
                    keywords: DoubleKeyword::InOut,
                },
            )
        } else {
            (
                "go",
                "to",
                node.child(0)?,
                Self {
                    keywords: DoubleKeyword::GoTo,
                },
            )
        };
        // Verify that the node is 'in' / 'go'
        if first_child.to_text(src.source_text())?.to_lowercase() != first {
            return None;
        }
        // Check if immediate sibling is 'out'/'to'
        let sibling = first_child.next_sibling()?;
        if sibling.to_text(src.source_text())?.to_lowercase() == second {
            let start = TextSize::try_from(first_child.start_byte()).unwrap();
            let end = TextSize::try_from(sibling.end_byte()).unwrap();
            let fix_start = TextSize::try_from(first_child.end_byte()).unwrap();
            let fix_end = TextSize::try_from(sibling.start_byte()).unwrap();
            let fix = Fix::safe_edit(Edit::deletion(fix_start, fix_end));
            return some_vec!(Diagnostic::new(violation, TextRange::new(start, end)).with_fix(fix));
        }
        // Can't fix this case, it may contain line continuations, comments, etc.
        some_vec!(Diagnostic::from_node(violation, &first_child))
    }

    fn entrypoints() -> Vec<&'static str> {
        vec![
            "inout",
            "keyword_statement", // goto
        ]
    }
}

pub(crate) mod settings {
    use crate::display_settings;
    use ruff_macros::CacheKey;
    use std::fmt::{Display, Formatter};

    #[derive(Debug, Clone, Default, CacheKey)]
    pub struct Settings {
        pub inout_with_space: bool,
        pub goto_with_space: bool,
    }

    impl Display for Settings {
        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
            display_settings! {
                formatter = f,
                namespace = "check.keyword-whitespace",
                fields = [self.inout_with_space, self.goto_with_space]
            }
            Ok(())
        }
    }
}
