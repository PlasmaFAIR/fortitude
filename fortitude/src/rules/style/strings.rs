use ruff_diagnostics::{Diagnostic, Edit, Fix, FixAvailability, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use ruff_text_size::TextSize;
use tree_sitter::Node;

use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};

/// ## What does it do?
/// Catches use of single-quoted strings.
///
/// ## Why is this bad?
/// For consistency, all strings should be either single-quoted or double-quoted.
/// Here, we enforce the use of double-quoted strings as this is the most common
/// style in other languages.
///
/// An exception is made for single-quoted strings that contain a `'"'` character,
/// as this is the preferred way to include double quotes in a string.
///
/// Fixes are not available for single-quoted strings containing escaped single
/// quotes (`"''"`).
#[derive(ViolationMetadata)]
pub(crate) struct SingleQuoteString {
    contains_escaped_quotes: bool,
}

impl Violation for SingleQuoteString {
    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;

    #[derive_message_formats]
    fn message(&self) -> String {
        "Character string uses single quotes".to_string()
    }

    fn fix_title(&self) -> Option<String> {
        if self.contains_escaped_quotes {
            return None;
        }
        Some("Replace with '\"'".to_string())
    }
}

impl AstRule for SingleQuoteString {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let text = node.to_text(src.source_text())?;
        if text.starts_with("'") && text.ends_with("'") && !text.contains('"') {
            // Search for occurrence of escaped single quotes within the string.
            // These are double single quotes, e.g. "''"
            if text.contains("''") && text.len() > 2 {
                return Some(vec![Diagnostic::from_node(
                    Self {
                        contains_escaped_quotes: true,
                    },
                    node,
                )]);
            }

            let start_byte = node.start_textsize();
            let end_byte = node.end_textsize();
            let edit_start =
                Edit::replacement("\"".to_string(), start_byte, start_byte + TextSize::from(1));
            let edit_end =
                Edit::replacement("\"".to_string(), end_byte - TextSize::from(1), end_byte);
            return some_vec!(Diagnostic::from_node(
                Self {
                    contains_escaped_quotes: false,
                },
                node,
            )
            .with_fix(Fix::safe_edits(edit_start, [edit_end])));
        }
        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["string_literal"]
    }
}
