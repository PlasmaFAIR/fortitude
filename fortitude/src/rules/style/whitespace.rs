/// Defines rules that enforce widely accepted whitespace rules.
use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::{SourceFile, UniversalNewlines};
use ruff_text_size::{TextLen, TextRange, TextSize};
use tree_sitter::Node;

use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode, TextRule};

/// ## What does it do?
/// Checks for tailing whitespace
///
/// ## Why is this bad?
/// Trailing whitespace is difficult to spot, and as some editors will remove it
/// automatically while others leave it, it can cause unwanted 'diff noise' in
/// shared projects.
#[derive(ViolationMetadata)]
pub(crate) struct TrailingWhitespace {}

impl AlwaysFixableViolation for TrailingWhitespace {
    #[derive_message_formats]
    fn message(&self) -> String {
        "trailing whitespace".to_string()
    }

    fn fix_title(&self) -> String {
        "Remove trailing whitespace".to_string()
    }
}

impl TextRule for TrailingWhitespace {
    fn check(_settings: &Settings, source_file: &SourceFile) -> Vec<Diagnostic> {
        let source = source_file.to_source_code();
        let mut violations = Vec::new();
        for line in source.text().universal_newlines() {
            let whitespace_bytes: TextSize = line
                .chars()
                .rev()
                .take_while(|c| c.is_whitespace())
                .map(TextLen::text_len)
                .sum();
            if whitespace_bytes > 0.into() {
                let range = TextRange::new(line.end() - whitespace_bytes, line.end());
                let edit = Edit::range_deletion(range);
                violations.push(Diagnostic::new(Self {}, range).with_fix(Fix::safe_edit(edit)));
            }
        }
        violations
    }
}

/// ## What does it do?
/// Checks for inline comments that aren't preceded by at least two spaces.
///
/// ## Why is this bad?
/// Inline comments that aren't separated from code by any whitespace can make
/// code hard to read. Other language style guides recommend the use of two
/// spaces before inline comments, so we recommend the same here.
///
/// ## References
/// - [PEP8 Python Style Guide](https://peps.python.org/pep-0008/)
/// - [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html#Horizontal_Whitespace)
#[derive(ViolationMetadata)]
pub(crate) struct IncorrectSpaceBeforeComment {}

impl AlwaysFixableViolation for IncorrectSpaceBeforeComment {
    #[derive_message_formats]
    fn message(&self) -> String {
        "need at least 2 spaces before inline comment".to_string()
    }

    fn fix_title(&self) -> String {
        "add extra whitespace".to_string()
    }
}
impl AstRule for IncorrectSpaceBeforeComment {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let source = src.to_source_code();
        let comment_start = node.start_textsize();
        // Get the line up to the start of the comment
        let line_index = source.line_index(comment_start);
        let line_start = source.line_start(line_index);
        let range = TextRange::new(line_start, comment_start);
        let line = source.slice(range);
        // Count whitespace characters at the end of the line
        let whitespace = line.chars().rev().take_while(|c| c.is_whitespace()).count();
        // If the line is empty or just filled with whitespace, exit
        if line.len() == whitespace {
            return None;
        }
        if whitespace < 2 {
            let edit = Edit::insertion("  "[whitespace..].to_string(), comment_start);
            return some_vec!(Diagnostic::from_node(Self {}, node).with_fix(Fix::safe_edit(edit)));
        }
        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["comment"]
    }
}
