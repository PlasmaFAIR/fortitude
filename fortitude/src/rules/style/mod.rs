pub mod double_colon_in_decl;
pub mod end_statements;
pub mod file_contents;
pub mod file_extensions;
pub mod functions;
pub mod implicit_none;
pub mod keywords;
pub mod line_length;
pub mod semicolons;
pub mod whitespace;

#[cfg(test)]
mod tests {
    use std::convert::AsRef;
    use std::path::Path;

    use anyhow::Result;
    use insta::assert_snapshot;
    use test_case::test_case;

    use crate::apply_common_filters;
    use crate::registry::Rule;
    use crate::rules::style::keywords;
    use crate::settings::{CheckSettings, Settings};
    use crate::test::test_path;

    #[test_case(Rule::LineTooLong, Path::new("S001.f90"))]
    #[test_case(Rule::UnnamedEndStatement, Path::new("S061.f90"))]
    #[test_case(Rule::MissingDoubleColon, Path::new("S071.f90"))]
    #[test_case(Rule::SuperfluousSemicolon, Path::new("S081.f90"))]
    #[test_case(Rule::MultipleStatementsPerLine, Path::new("S082.f90"))]
    #[test_case(Rule::TrailingWhitespace, Path::new("S101.f90"))]
    #[test_case(Rule::IncorrectSpaceBeforeComment, Path::new("S102.f90"))]
    #[test_case(Rule::SuperfluousImplicitNone, Path::new("S201.f90"))]
    #[test_case(Rule::MultipleModules, Path::new("S211.f90"))]
    #[test_case(Rule::ProgramWithModule, Path::new("S212.f90"))]
    #[test_case(Rule::FunctionMissingResult, Path::new("S221.f90"))]
    #[test_case(Rule::KeywordsMissingSpace, Path::new("S231.f90"))]
    #[test_case(Rule::KeywordHasWhitespace, Path::new("S231.f90"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("style").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }

    #[test_case(Rule::SuperfluousImplicitNone, Path::new("S201_ok.f90"))]
    fn rules_pass(rule_code: Rule, path: &Path) -> Result<()> {
        let diagnostics = test_path(
            Path::new("style").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        assert!(
            diagnostics.is_empty(),
            "Test source has no warnings, but some were raised:\n{diagnostics}"
        );
        Ok(())
    }

    #[test_case(Rule::LineTooLong, Path::new("S001_line_length_20.f90"))]
    fn line_too_long_line_length_20(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());

        let default = Settings::default();
        #[allow(clippy::needless_update)]
        let settings = Settings {
            check: CheckSettings {
                line_length: 20,
                ..default.check
            },
            ..default
        };
        let diagnostics = test_path(
            Path::new("style").join(path).as_path(),
            &[rule_code],
            &settings,
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }

    #[test_case(Rule::KeywordsMissingSpace, Path::new("S231.f90"))]
    #[test_case(Rule::KeywordHasWhitespace, Path::new("S231.f90"))]
    fn keyword_whitespace_include_inout_goto(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!(
            "{}_{}_include_inout_goto",
            rule_code.as_ref(),
            path.to_string_lossy()
        );

        let default = Settings::default();
        #[allow(clippy::needless_update)]
        let settings = Settings {
            check: CheckSettings {
                keyword_whitespace: keywords::settings::Settings {
                    inout_with_space: true,
                    goto_with_space: true,
                },
                ..default.check
            },
            ..default
        };
        let diagnostics = test_path(
            Path::new("style").join(path).as_path(),
            &[rule_code],
            &settings,
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }
}
