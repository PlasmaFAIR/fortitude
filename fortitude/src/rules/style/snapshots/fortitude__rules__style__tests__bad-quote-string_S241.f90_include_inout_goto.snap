---
source: fortitude/src/rules/style/mod.rs
expression: diagnostics
snapshot_kind: text
---
./resources/test/fixtures/style/S241.f90:4:12: S241 [*] String uses double quotes but single quotes preferred
  |
2 |   implicit none (type, external)
3 |
4 |   print *, "Hello, World!"
  |            ^^^^^^^^^^^^^^^ S241
5 |   print *, 'Hello, World!'
6 |   print *, 'Hello, "World"!'
  |
  = help: Replace double quotes with single quotes

ℹ Safe fix
1 1 | program p
2 2 |   implicit none (type, external)
3 3 | 
4   |-  print *, "Hello, World!"
5 4 |   print *, 'Hello, World!'
  5 |+  print *, 'Hello, World!'
6 6 |   print *, 'Hello, "World"!'
7 7 |   print *, "Hello, ""World""!"
8 8 |   print *, 'Hello, ''World''!'

./resources/test/fixtures/style/S241.f90:7:12: S241 String uses double quotes but single quotes preferred
  |
5 |   print *, 'Hello, World!'
6 |   print *, 'Hello, "World"!'
7 |   print *, "Hello, ""World""!"
  |            ^^^^^^^^^^^^^^^^^^^ S241
8 |   print *, 'Hello, ''World''!'
9 |   print *, "Hello, &
  |

./resources/test/fixtures/style/S241.f90:9:12: S241 [*] String uses double quotes but single quotes preferred
   |
 7 |     print *, "Hello, ""World""!"
 8 |     print *, 'Hello, ''World''!'
 9 |     print *, "Hello, &
   |  ____________^
10 | |             & World!"
   | |_____________________^ S241
11 |     print *, 'Hello, &
12 |               & World!'
   |
   = help: Replace double quotes with single quotes

ℹ Safe fix
6  6  |   print *, 'Hello, "World"!'
7  7  |   print *, "Hello, ""World""!"
8  8  |   print *, 'Hello, ''World''!'
9     |-  print *, "Hello, &
10    |-            & World!"
11 9  |   print *, 'Hello, &
12 10 |             & World!'
13 11 |   print *, 'Hello, &
   12 |+            & World!'
   13 |+  print *, 'Hello, &
14 14 |             & "World"!'
15 15 |   print *, "Hello, &
16 16 |             & ""World""!"

./resources/test/fixtures/style/S241.f90:15:12: S241 String uses double quotes but single quotes preferred
   |
13 |     print *, 'Hello, &
14 |               & "World"!'
15 |     print *, "Hello, &
   |  ____________^
16 | |             & ""World""!"
   | |_________________________^ S241
17 |     print *, 'Hello, &
18 |               & ''World''!'
   |
